### PR TITLE
Melee Combat Events

### PR DESCRIPTION
**Pull request by OptimShi**
Added handlers for the following melee combat events:
ATTACKER_NOTIFICATION_EVENT
DEFENDER_NOTIFICATION_EVENT
EVASION_ATTACKER_NOTIFICATION_EVENT
EVASION_DEFENDER_NOTIFICATION_EVENT

### CODE REVIEW

File: aclogview/CM_Combat.cs [hunk 1]
```diff
@@ -161,4 +185,121 @@ public class QueryHealthResponse : Message {
...
+        public ulong attack_conditions;
+
+        public static AttackerNotificationEvent read(BinaryReader binaryReader)
+        {
+            AttackerNotificationEvent newObj = new AttackerNotificationEvent();
+            newObj.defenders_name = PStringChar.read(binaryReader);
+            newObj.damage_type = binaryReader.ReadUInt32();
+            newObj.severity = binaryReader.ReadDouble();
+            newObj.damage = binaryReader.ReadUInt32();
+            newObj.critical = binaryReader.ReadBoolean();
```
Comment (tfarley) [inline aclogview/CM_Combat.cs]: From what I can see, it looks like critical is actually a full int, not just a single byte.

File: aclogview/CM_Combat.cs [hunk 2]
```diff
@@ -161,4 +185,121 @@ public class QueryHealthResponse : Message {
...
+
+        public static DefenderNotificationEvent read(BinaryReader binaryReader)
+        {
+            DefenderNotificationEvent newObj = new DefenderNotificationEvent();
+            newObj.defenders_name = PStringChar.read(binaryReader);
+            newObj.damage_type = binaryReader.ReadUInt32();
+            newObj.severity = binaryReader.ReadDouble();
+            newObj.damage = binaryReader.ReadUInt32();
+            newObj.part = binaryReader.ReadUInt32();
+            newObj.critical = binaryReader.ReadBoolean();
```
Comment (tfarley) [inline aclogview/CM_Combat.cs]: Same, full int instead of bool for critical.

File: aclogview/CM_Combat.cs [hunk 3]
```diff
@@ -161,4 +185,121 @@ public class QueryHealthResponse : Message {
...
+            rootNode.Nodes.Add("damage = " + damage);
+            rootNode.Nodes.Add("critical = " + critical);
+            rootNode.Nodes.Add("attackConditions = " + attack_conditions);
+            treeView.Nodes.Add(rootNode);
+        }
+    }
+
+    public class DefenderNotificationEvent : Message
+    {
+        public PStringChar defenders_name;
```
Comment (tfarley) [inline aclogview/CM_Combat.cs]: This should be attackers_name

File: aclogview/CM_Combat.cs [hunk 4]
```diff
@@ -161,4 +185,121 @@ public class QueryHealthResponse : Message {
...
+            newObj.attack_conditions = binaryReader.ReadUInt64();
+            Util.readToAlign(binaryReader);
+            return newObj;
+        }
+
+        public override void contributeToTreeView(TreeView treeView)
+        {
+            TreeNode rootNode = new TreeNode(this.GetType().Name);
+            rootNode.Expand();
+            rootNode.Nodes.Add("defenders_name = " + defenders_name);
```
Comment (tfarley) [inline aclogview/CM_Combat.cs]: attackers_name

File: aclogview/CM_Combat.cs [hunk 5]
```diff
@@ -161,4 +185,121 @@ public class QueryHealthResponse : Message {
...
+            rootNode.Nodes.Add("part = " + (BodyPart)part);
+            rootNode.Nodes.Add("critical = " + critical);
+            rootNode.Nodes.Add("attackConditions = " + attack_conditions);
+            treeView.Nodes.Add(rootNode);
+        }
+    }
+
+    public class EvasionAttackerNotificationEvent : Message
+    {
+        public PStringChar attacker;
```
Comment (tfarley) [inline aclogview/CM_Combat.cs]: Please make this defenders_name, just to be consistent

File: aclogview/CM_Combat.cs [hunk 6]
```diff
@@ -161,4 +185,121 @@ public class QueryHealthResponse : Message {
...
+            newObj.attacker = PStringChar.read(binaryReader);
+            Util.readToAlign(binaryReader);
+            return newObj;
+        }
+
+        public override void contributeToTreeView(TreeView treeView)
+        {
+            TreeNode rootNode = new TreeNode(this.GetType().Name);
+            rootNode.Expand();
+            rootNode.Nodes.Add("attacker = " + attacker);
```
Comment (tfarley) [inline aclogview/CM_Combat.cs]: defenders_name

File: aclogview/CM_Combat.cs [hunk 7]
```diff
@@ -161,4 +185,121 @@ public class QueryHealthResponse : Message {
...
+            treeView.Nodes.Add(rootNode);
+        }
+    }
+
+    public class EvasionDefenderNotificationEvent : Message
+    {
+        public PStringChar attacker;
```
Comment (tfarley) [inline aclogview/CM_Combat.cs]: Please make this attackers_name, just to be consistent

File: aclogview/CM_Combat.cs [hunk 8]
```diff
@@ -161,4 +185,121 @@ public class QueryHealthResponse : Message {
...
+            newObj.attacker = PStringChar.read(binaryReader);
+            Util.readToAlign(binaryReader);
+            return newObj;
+        }
+
+        public override void contributeToTreeView(TreeView treeView)
+        {
+            TreeNode rootNode = new TreeNode(this.GetType().Name);
+            rootNode.Expand();
+            rootNode.Nodes.Add("defender = " + attacker);
```
Comment (tfarley) [inline aclogview/CM_Combat.cs]: attackers_name

File: aclogview/Enums/Combat.cs [hunk 9]
```diff
@@ -111,3 +111,36 @@ public enum PowerBarMode {
     PBM_JUMP,
     PBM_DDD
 }
+
+public enum BodyPart
+{
+    UNDEFINED,
```
Comment (tfarley) [inline aclogview/Enums/Combat.cs]: These actually all need to shift by -1, so UNDEFINED = -1, HEAD = 0, CHEST = 1, etc. I would suggest just getting rid of the UNDEFINED entry.

File: aclogview/Enums/Combat.cs [hunk 10]
```diff
@@ -111,3 +111,36 @@ public enum PowerBarMode {
...
+    HAND,
+    UPPER_LEG,
+    LOWER_LEG,
+    FOOT,
+    HORN,
+    FRONT_LEG,
+    SKIP, // Skip 12
+    FRONT_FOOT,
+    REAR_LEG,
+    SKIP_2, // Skip 15
```
Comment (tfarley) [inline aclogview/Enums/Combat.cs]: After the above change, these two comments then become Skip 11 and Skip 14

**Review by tfarley — CHANGES_REQUESTED** [review]
Thanks for the PR. Just left a few comments for small fixes.

**Comment by OptimShi** [comment]
Thanks for taking the time to look at this and catch my mistakes! I've adjusted with your feedback and also added the enum for the "attack_conditions" bitfield (not currently reading them, but they are least there for reference for the time being)

**Pull request merged**